src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,19 +2205,22 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
-                {
+                var index = 0;
+                var count = this.Count;
+                var removed = 0;
+                for(;;)
+                {
+                    if (index + removed >= count)
+                        break;
+                    var item = result[index];
```

Comment:
In my evaluation, I was referring to best-case performance where *n* is the size of the list and the predicate is `x => false`. You would have a separate input to this *m* which is the number of items actually removed, with the original complexity being close to O(n log m) - best case O(n), worst case something close to O(n log n). Since the increase in *m* directly causes a reduction in the size of the list, someone would have to remind me what the worst case impact actually is (O(n log n) is an upper bound).